[Keymap][user] user dvorak_42_key ergodox_ez and crkbd keymaps

luc-languagetools opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->
<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description
Add user dvorak_42_key in userspace. add crkbd keymap.
<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [X] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

waffle87 commented on keyboards/crkbd/keymaps/dvorak_42_key/keymap.c:
```diff
@@ -0,0 +1,287 @@
+
```
All source files need a GPL2+ license header:
```
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

luc-languagetools replied:
done, thank you.

drashna commented on keyboards/crkbd/keymaps/dvorak_42_key/config.h:
```diff
@@ -0,0 +1,45 @@
+
+#pragma once
+
+//#define USE_MATRIX_I2C
```
```suggestion
```

drashna commented on keyboards/crkbd/keymaps/dvorak_42_key/keymap.c:
```diff
@@ -0,0 +1,290 @@
+#include QMK_KEYBOARD_H
+#include <stdio.h>
+#include "dvorak_42_key.h"
+#include "rgblight.h"
```
Shouldn't be needed:
```suggestion
```

drashna commented on keyboards/crkbd/keymaps/dvorak_42_key/keymap.c:
```diff
@@ -0,0 +1,290 @@
+  
+
+#define DISPLAY_LAYER_NAME(LAYER_NAME, LAYER_STRING) \
+  if(biton32(layer_state) == LAYER_NAME) { \
```
```suggestion
  if(get_highest_layer(layer_state) == LAYER_NAME) { \
```

drashna commented on users/dvorak_42_key/config.h:
```diff
@@ -0,0 +1,6 @@
+// Copyright 2022 LucW (@luc-languagetools)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#define DOUBLE_TAP_SHIFT_TURNS_ON_CAPS_WORD
```
```suggestion
#pragma once

#define DOUBLE_TAP_SHIFT_TURNS_ON_CAPS_WORD
```

luc-languagetools commented:
note: the latest CI build error mentions the firmware is too big:
```
Checking file size of crkbd_r2g_dvorak_42_key.hex
 * The firmware is too large! 28918/28672 (246 bytes over)
 [ERRORS]
```
however when I build locally, I get the following result:
```
The firmware size is fine - 25776/28672 (89%, 2896 bytes free)
```

waffle87 commented:
> note: the latest CI build error mentions the firmware is too big:

I assume the CI uses avr-gcc 5.x, which results in larger binaries, while you probably have 8.x. You can see if there's anything that can be trimmed — [squeezing AVR](https://docs.qmk.fm/#/squeezing_avr), but I'd say it isn't too big of a deal.

fauxpark commented on users/dvorak_42_key/dvorak_42_key.h:
```diff
@@ -0,0 +1,202 @@
+#define MACRO_SCREEN_NUM(MACRO_NAME,NUM) \
+        case MACRO_NAME:\
+             if (record->event.pressed) {\
+                return MACRO( D(LCTL), T(A), U(LCTL), T(NUM), END);\
+            }\
+        break;\
+
+#define MACRO_SCREEN_REG(MACRO_NAME,NUM) \
+        case MACRO_NAME:\
+             if (record->event.pressed) {\
+                return MACRO( D(LCTL), T(A), U(LCTL), D(LSFT), T(SCOLON), U(LSFT),\
+                              T(R),\
+                              T(E),\
+                              T(A),\
+                              T(D),\
+                              T(R),\
+                              T(E),\
+							  T(G),\
+                              T(SPC),\
+							  T(NUM),\
+							  T(ENTER),\
+                             END);\
+            }\
+        break;\
+
+#define MACRO_SCREEN_PASTE(MACRO_NAME,NUM)\
+        case MACRO_NAME:\
+             if (record->event.pressed) {\
+                return MACRO( D(LCTL), T(A), U(LCTL), D(LSFT), T(SCOLON), U(LSFT),\
+                              T(P),\
+                              T(A),\
+                              T(S),\
+                              T(T),\
+                              T(E),\
+                              T(SPC),\
+							  T(NUM),\
+							  T(ENTER),\
+                             END);\
+            }\
+        break;\
```
```suggestion
```
These are for the old macro system which has been removed and will not work.

fauxpark commented on users/dvorak_42_key/dvorak_42_key.c:
```diff
@@ -0,0 +1,230 @@
+        case SHELL_LS:
+            SEND_STRING("ls\n");
+            return true;
+            break;
```
`break` after `return` is dead code.

fauxpark commented on users/dvorak_42_key/dvorak_42_key.h:
```diff
@@ -0,0 +1,202 @@
+
+enum custom_keycodes {
+  PLACEHOLDER = SAFE_RANGE, // can always be here
+  EPRM,
+  VRSN,
+  RGB_SLD,
```
```suggestion
```

fauxpark reviewed (changes requested):
Nope, CI uses AVR-GCC 8. The problem is likely due to using a lot of rather heavy features (tap dance, caps word, RGB, OLED, etc), and a fair amount of layers. So you may have to remove some stuff.

luc-languagetools commented:
I notice that the "rev1" build succeeds but not the r2g. Is there a way to declare that my keymap only builds on rev1?

fauxpark commented:
You can move your keymap folder under the rev1 keymaps folder (create it if it doesn't exist).

luc-languagetools commented:
@drashna thank you for your time. i've implemented your changes and i've moved the keymap to `keyboards/crkbd/rev1/keymaps/dvorak_42_key`. Does this change now meet the pre-requisite to be merged ?

fauxpark commented on users/dvorak_42_key/dvorak_42_key.c:
```diff
@@ -0,0 +1,217 @@
+            break;            
+
+        case WINDOWS10_WORKSPACE_LEFT:
+            SEND_STRING(SS_LGUI(SS_LCTL(SS_TAP(X_LEFT))));
```
```suggestion
            tap_code16(C(G(KC_LEFT)));
```

fauxpark commented on users/dvorak_42_key/dvorak_42_key.c:
```diff
@@ -0,0 +1,217 @@
+            SEND_STRING(SS_LGUI(SS_LCTL(SS_TAP(X_LEFT))));
+            break;
+        case WINDOWS10_WORKSPACE_RIGHT:
+            SEND_STRING(SS_LGUI(SS_LCTL(SS_TAP(X_RIGHT))));
```
```suggestion
            tap_code16(C(G(KC_RIGHT)));
```

fauxpark commented on users/dvorak_42_key/dvorak_42_key.c:
```diff
@@ -0,0 +1,217 @@
+            SEND_STRING(SS_LGUI(SS_LCTL(SS_TAP(X_RIGHT))));
+            break;
+        case WINDOWS10_TASK_VIEW:
+            SEND_STRING(SS_LGUI(SS_TAP(X_TAB)));
```
```suggestion
            tap_code16(G(KC_TAB));
```

fauxpark commented on users/dvorak_42_key/dvorak_42_key.c:
```diff
@@ -0,0 +1,217 @@
+            SEND_STRING(SS_LCTL("a") "d");
+            break;
+        case SCREEN_RENAME:
+            SEND_STRING(SS_LCTL("a") SS_LSFT("a"));
```
```suggestion
            SEND_STRING(SS_LCTL("a") "A");
```

fauxpark commented on users/dvorak_42_key/dvorak_42_key.c:
```diff
@@ -0,0 +1,217 @@
+            SEND_STRING(SS_LCTL("a") "]");
+            break;
+        case SCREEN_KILL:
+            SEND_STRING(SS_LCTL("a") "k" "y");
```
```suggestion
            SEND_STRING(SS_LCTL("a") "ky");
```

fauxpark commented on users/dvorak_42_key/dvorak_42_key.c:
```diff
@@ -0,0 +1,217 @@
+            SEND_STRING(SS_LCTL("a") "k" "y");
+            break;            
+        case SCREEN_READREG_1:
+            tap_code16(C(KC_A));
+            tap_code16(S(KC_SEMICOLON));
+            SEND_STRING("readreg 1\n");
```
```suggestion
            SEND_STRING(SS_LCTL("a") ":readreg 1\n");
```

fauxpark commented on users/dvorak_42_key/dvorak_42_key.c:
```diff
@@ -0,0 +1,217 @@
+            SEND_STRING("readreg 1\n");
+            break;
+        case SCREEN_READREG_2:
+            tap_code16(C(KC_A));
+            tap_code16(S(KC_SEMICOLON));
+            SEND_STRING("readreg 2\n");
```
```suggestion
            SEND_STRING(SS_LCTL("a") ":readreg 2\n");
```

fauxpark commented on users/dvorak_42_key/dvorak_42_key.c:
```diff
@@ -0,0 +1,217 @@
+            SEND_STRING("readreg 2\n");
+            break;
+        case SCREEN_READREG_3:
+            tap_code16(C(KC_A));
+            tap_code16(S(KC_SEMICOLON));
+            SEND_STRING("readreg 3\n");
```
```suggestion
            SEND_STRING(SS_LCTL("a") ":readreg 3\n");
```

fauxpark commented on users/dvorak_42_key/dvorak_42_key.c:
```diff
@@ -0,0 +1,217 @@
+            break;
+
+        case SCREEN_PASTEREG_1:
+            tap_code16(C(KC_A));
+            tap_code16(S(KC_SEMICOLON));
+            SEND_STRING("paste 1\n");
```
```suggestion
            SEND_STRING(SS_LCTL("a") ":paste 1\n");
```

fauxpark commented on users/dvorak_42_key/dvorak_42_key.c:
```diff
@@ -0,0 +1,217 @@
+            SEND_STRING("paste 1\n");
+            break;
+        case SCREEN_PASTEREG_2:
+            tap_code16(C(KC_A));
+            tap_code16(S(KC_SEMICOLON));
+            SEND_STRING("paste 2\n");
```
```suggestion
            SEND_STRING(SS_LCTL("a") ":paste 2\n");
```

fauxpark commented on users/dvorak_42_key/dvorak_42_key.c:
```diff
@@ -0,0 +1,217 @@
+            SEND_STRING("paste 2\n");
+            break;
+        case SCREEN_PASTEREG_3:
+            tap_code16(C(KC_A));
+            tap_code16(S(KC_SEMICOLON));
+            SEND_STRING("paste 3\n");
```
```suggestion
            SEND_STRING(SS_LCTL("a") ":paste 3\n");
```

luc-languagetools commented:
@drashna can this change be merged ? I believe i've implemented all of the requested changes.

luc-languagetools commented:
Hi, can this be merged ? Please let me know. Thank you.

luc-languagetools commented:
Thank you !
